getlatest.go

```diff
@@ -0,0 +1,135 @@
+package namesys
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	ggio "github.com/gogo/protobuf/io"
+	"github.com/gogo/protobuf/proto"
+
+	"github.com/libp2p/go-libp2p-core/helpers"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	pb "github.com/libp2p/go-libp2p-pubsub-router/pb"
+)
+
+type getLatestProtocol struct {
+	ctx  context.Context
+	host host.Host
+}
+
+func newGetLatestProtocol(ctx context.Context, host host.Host, getLocal func(key string) ([]byte, error)) *getLatestProtocol {
+	p := &getLatestProtocol{ctx, host}
+
+	host.SetStreamHandler(PSGetLatestProto, func(s network.Stream) {
+		p.receive(s, getLocal)
+	})
+
+	return p
+}
+
+func (p *getLatestProtocol) receive(s network.Stream, getLocal func(key string) ([]byte, error)) {
+	defer helpers.FullClose(s)
+
+	msg := &pb.RequestLatest{}
+	if err := readMsg(p.ctx, s, msg); err != nil {
+		log.Infof("error reading request from %s: %s", s.Conn().RemotePeer(), err)
+		s.Reset()
+		return
+	}
+
+	response, err := getLocal(msg.Identifier)
+	var respProto pb.RespondLatest
+
+	if err != nil {
+		respProto = pb.RespondLatest{Status: pb.RespondLatest_NOT_FOUND}
+	} else {
+		respProto = pb.RespondLatest{Data: response}
+	}
+
+	if err := writeMsg(p.ctx, s, &respProto); err != nil {
+		return
+	}
+}
+
+func (p getLatestProtocol) Get(ctx context.Context, pid peer.ID, key string) ([]byte, error) {
+	peerCtx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+
+	s, err := p.host.NewStream(peerCtx, pid, PSGetLatestProto)
+	if err != nil {
+		return nil, err
+	}
+	defer helpers.FullClose(s)
+
+	msg := &pb.RequestLatest{Identifier: key}
+
+	if err := writeMsg(ctx, s, msg); err != nil {
+		return nil, err
+	}
+	s.Close()
+
+	response := &pb.RespondLatest{}
+	if err := readMsg(ctx, s, response); err != nil {
+		return nil, err
+	}
+
+	switch response.Status {
+	case pb.RespondLatest_SUCCESS:
+		return response.Data, nil
+	case pb.RespondLatest_NOT_FOUND:
+		return nil, nil
+	default:
+		return nil, errors.New("get-latest: received unknown status code")
+	}
+}
+
+func writeMsg(ctx context.Context, s network.Stream, msg proto.Message) error {
```

Comment:
why the dance with the goroutine?

Reply:
without the "dance" there's no way to do a stream send/receive that is cancellable. Alternative discussed above (https://github.com/libp2p/go-libp2p-pubsub-router/pull/33#discussion_r312343726) was to use Deadline with some arbitrary small duration.

Any better ideas?

Reply:
why do we care about cancellation here?

Reply:
A stream deadline will be simpler.

Reply:
> why do we care about cancellation here?

Because the user can cancel their contexts and shutdown the subscription? Technically this will also give an "incorrect" result if a user gets a background update after the subscription has been cancelled. However, it's likely more about swiftly freeing up streams (and connections) for cancellation.

If we'd prefer to just slap a 5 second timeout on as I originally did (https://github.com/libp2p/go-libp2p-pubsub-router/pull/33#discussion_r312343726), that's probably ok (although setting arbitrary times in the code feels not great, what if your connects are really slow).

@Stebalien @vyzo thoughts?

Reply:
Even if the user cancels the read/write goroutines will linger in i/o since it is done without deadline.
I am arguing that it is not very useful functionality to have cancellation here.

Reply:
> Even if the user cancels the read/write goroutines will linger in i/o since it is done without deadline.

When we cancel then we send a stream reset. Is that not enough to kill the IO?

Reply:
After some slack messages with @vyzo to clarify this is roughly where we're at (@vyzo please correct me if I've misrepresented anything).

Deadlines: 
 - Cleaner in code
 - No extra goroutines in the nice case where everything finishes quickly with no errors
 - With a 1 min timeout shouldn't leave us too many extraneous streams + goroutines even under error/cancellation

Cancellation:
- Makes more sense in programmatic logic since we already have and use contexts
- Doesn't rely on arbitrary numbers (like 1 minute)
- Adds an extra goroutine (it's currently two goroutines, one for read and another for write, but if needed we could combine them) for the duration of the request, even in the good case
- Frees up the goroutine on cancellation without waiting a minute
  - potentially very useful if the Fetch protocol is used beyond IPNS where people might use it like bitswap and send out many Fetches and cancel them

Going with either option at this point is fine and is fairly low programming effort given that both have been implemented over the course of this PR. As long as we document why we're choosing the one we go with (and we get this merged soon) I'm happy.

@Stebalien any thoughts?

Reply:
Hi all! I believe this is the only decision point holding up this PR. @Stebalien @vyzo would you mind weighing on the side of either `Deadlines` or `Cancellation`, or explicitly letting us know that you don't have a strong opinion, and that @aschmahmann should use his best judgement? _Thank you!_

Reply:
Let's leave the goroutine. Goroutines in this case aren't going to be a huge issue and we do stuff like this all the time (secio, dht, etc.).

Reply:
ok, fine.